### ISSUE TITLE
Correlation between training data for Q value used for training and result obtained from playouts at Temp = 0

### ISSUE THREAD

**Issue by amjshl** [LeelaChessZero/lc0#330]
This was due to the discussions regarding quality of training data from games used to train the neural network.

I ran a series of experiments beginning with the positions generated by playing 10 games using 800 playouts and Temp = 1 (test id: 10971). This generated around 1367 positions. Starting from each of this positions, I ran lc0 with different playouts but with Temp = 0, till the end of game and used the end result for calculating the mean squared error against each other, tabulated as a cross table. To get a perspective on the mse values, I added a column with random result predictor where the result was randomly generated while preserving the ratio of win/loss/draw. Here is the result.

![screenshot_2018-09-01_22-55-29](https://user-images.githubusercontent.com/42904180/44952755-69bc0280-ae3c-11e8-9f7a-3f5608ab2e8a.png)

The above cross table only includes positions which had an evaluation between -0.95 and +0.95 for its principle variation (with the case of 800 playouts, Temp=1).

This indicates  that the training data can be significantly improved by even using playouts=1 or 10, temp = 0 from each of the positions. Assuming average game length is 100, this would amount to 50*10 = 500 playouts per position which has about 70% impact on game generation rate. This way we get the variety in position with T=1 but with better accuracy (of T=0) on the end result.

For those interested to know more about the data, I retabulated the same data but divided it into different Threshold values for PV and different ply number of starting position, which also provide some interesting insights.

Different range of absolute value of the PV:

![screenshot_2018-09-01_23-29-21](https://user-images.githubusercontent.com/42904180/44952870-fcf63780-ae3e-11e8-8d41-ee10f8e11766.png)

Different range of ply number of the starting position:

![screenshot_2018-09-01_23-32-11](https://user-images.githubusercontent.com/42904180/44952879-45adf080-ae3f-11e8-9edb-e676efd0190a.png)

I believe using game results generated from T=0, even for small number of playouts will give rise to more accurate value network, which I think strongly correlates with the raw tactical strength of the network. MCTS corrects the evaluation function but requires more nodes the less accurate the evaluation function is. With accurate value, MCTS will converge to actual value faster.

**Comment by killerducky** [comment]
I'm having a hard time reading the tables. Is there an entry that compares the game result (not Q) of many games played from the same position with T=1 and T=0?

**Comment by amjshl** [comment]
I only played one game from every position, with T=0, playouts = 1, 10, 50, 200, and 800, and with T=1, playouts = 800. Why play multiple games from the same position since in training we only use result from just one game? 

Sorry if the table is a bit hard to read. Row with "Training data, result, T=1, 800 playouts" is the reference and is compared against, T=0, playouts = 1, 10, 50, 200, 800 in the same row.

**Comment by MaurizioDeLeo** [comment]
I think that the data presentation is not very clear. I refactored the table to be more clear.
What we are targeting is to have a result as similar as possible to the "800 playouts Temperature=0 result" (which is the best approximation of the "true" value of the position).
So the smaller the number in column "I", the better. The other comparisons don't add much information.

![table](https://user-images.githubusercontent.com/42923518/44966461-d3b2d580-af2a-11e8-8a50-985ce7a16248.jpg)

So the table suggest that averaging the result with the current Q is much better than the current approach of taking the Temperature=1 result. 
A mean square of 0.67 means that 80% of the time, the T=1 result is skewed of one "notch" versus the "true" result (win instead of draw or draw instead of loss).
Another possible approach is to complete the game in parallel at 200 playouts and T=0. However computationally this might be heavy (about 10 time increase in training time in my estimate).

A possible improvement of the data gathering would be to run at T=0 and much higher playouts (e.g. 10'000) to get a better estimate of the "true" value.

**Comment by amjshl** [comment]
@MaurizioDeLeo  Thanks for giving a much clearer summary than I could. That is precisely what I intended to convey. 

  Yes, my assumption is that the result of T=0 at 800 playouts gives the most accurate value but that is very computationally intensive. As a compromise, running just 1, 10, or 50 playouts with T=0 is cheaper to compute but gives a better prediction accuracy than T=1 and 800 playouts used during training games. 

 The challenge is that we still need T=1 to generate greater variety of positions and explore new moves, but use T=0 only for determining the value of a position, but not generate positions for training.

 What's interesting is Table number 2. The training target provided by (T=1 playouts = 800) is especially bad for positions with score (q value after search) between 0.2 and 0.6 (or -0.2 and -0.6 where black has advantage) compared to the more accurate value obtained by (T = 0 playouts = 800). Agreed, that there are fewer positions analyzed for that range in my experiments and thus statistical uncertainty of the mse is higher, but there does seems to be a signal in that direction. IMO most positions where conversion from draw to win, or from loss to draw, need accurate estimation of scores falling in that range.

  MCTS will eventually converge to the actual solution no matter how inaccurate the training data (even a random value generator will converge given sufficiently high number of playouts) , but my hypothesis is that it will converge much faster if the training target is more accurate.
